hathor/simulator/simulator.py

```diff
@@ -0,0 +1,118 @@
+        peer_id = PeerId()
+        manager = HathorManager(
+            self.clock,
+            peer_id=peer_id,
+            network=network,
+            wallet=wallet,
+        )
+
+        manager.reactor = self.clock
+        manager.test_mode = TestMode.DISABLED
+        manager.avg_time_between_blocks = 64
+        manager._full_verification = True
+        manager.start()
+        self.run_to_completion()
+
+        # Don't use it anywhere else. It is unsafe to generate mnemonic words like this.
+        # It should be used only for testing purposes.
+        m = Mnemonic('english')
+        words = m.to_mnemonic(bytes(random.randint(0, 255) for _ in range(32)))
+        wallet.unlock(words=words, tx_storage=manager.tx_storage)
+        return manager
+
+    def run_to_completion(self):
+        """ This will advance the test's clock until all calls scheduled are done.
+        """
+        for call in self.clock.getDelayedCalls():
+            amount = max(0, call.getTime() - self.clock.seconds())
+            self.clock.advance(amount)
+
+    def apply_patches(self):
+        from hathor.transaction import BaseTransaction
+
+        def verify_pow(self: BaseTransaction) -> None:
+            assert self.hash is not None
+
+        self._original_verify_pow = BaseTransaction.verify_pow
+        BaseTransaction.verify_pow = verify_pow
+
+    def remove_patches(self):
+        from hathor.transaction import BaseTransaction
+        BaseTransaction.verify_pow = self._original_verify_pow
+
+    def add_peer(self, name: str, peer: 'HathorManager') -> None:
+        if name in self.peers:
+            raise ValueError('Duplicate peer name')
+        self.peers[name] = peer
+
+    def get_peer(self, name: str) -> 'HathorManager':
+        return self.peers[name]
+
+    def add_connection(self, conn: 'FakeConnection') -> None:
+        self.connections.append(conn)
+
+    def set_seed(self, seed: int) -> None:
+        self.seed = seed
+        random.seed(self.seed)
+        numpy.random.seed(self.seed)
+
+    def run(self, interval: float, step: float = 0.25, status_interval: float = 60.0) -> None:
+        initial = self.clock.seconds()
+        latest_time = self.clock.seconds()
+        t0 = time.time()
+        while self.clock.seconds() <= initial + interval:
+            for conn in self.connections:
+                conn.run_one_step()
+            if self.clock.seconds() - latest_time >= status_interval:
+                t1 = time.time()
+                print('[{:8.2f}][rate={:8.2f}] t={:15.2f}    dt={:8.2f}    toBeRun={:8.2f}    delayedCall={}'.format(
+                    t1 - t0,
+                    (self.clock.seconds() - initial) / (t1 - t0),
+                    self.clock.seconds(),
+                    self.clock.seconds() - initial,
+                    interval - self.clock.seconds() + initial,
+                    len(self.clock.getDelayedCalls()),
+                ))
```

Comment:
When a .format gets this big it's really hard to understand what goes where. An alternative is using variables and an f-string like

```python
foo = ...
print(f'... {foo:fmt} ...')
```
What do you think?

Reply:
Really better. I also added comments explaining each variable. Can you review again, please?

Reply:
Looks great!